modMcf/src/org/aion/mcf/trie/TrieImpl.java

```diff
@@ -881,6 +882,64 @@ private void traceTrie(byte[] stateRoot, ScanAction action) {
     //        return true;
     //    }
 
+    @Override
+    public Map<ByteArrayWrapper, byte[]> getReferencedTrieNodes(byte[] value, int limit) {
+        CollectMappings collect = new CollectMappings();
+        ArrayList<byte[]> hashes = new ArrayList<>();
+        Value node = Value.fromRlpEncoded(value);
+
+        synchronized (cache) {
+            appendHashes(node, hashes);
+
+            while (!hashes.isEmpty() && collect.getSize() < limit) {
+                byte[] myHash = hashes.remove(0);
```

Comment:
just think about you could avoid the remove call. Considering use the following code snipet and add element during the for loop might be faster.
https://stackoverflow.com/questions/993025/adding-elements-to-a-collection-during-iteration
```
List<ZeObj> myList = new ArrayList<ZeObj>();
    // populate the list with whatever
            ........
    int noItems = myList.size();
    for (int i = 0; i < noItems; i++) {
        ZeObj currItem = myList.get(i);
        // when you want to add, simply add the new item at last and
        // increment the stop condition
        if (currItem.asksForMore()) {
            myList.add(new ZeObj());
            noItems++;
        }
    }
```